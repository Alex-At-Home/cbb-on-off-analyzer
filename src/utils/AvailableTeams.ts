
export type AvailableTeamMeta = {
  team: string,
  year: string,
  gender: string,
  index_template: string
}

export class AvailableTeams {

  private static flatten(
    arr: Array<Array<AvailableTeamMeta>>
  ): Array<AvailableTeamMeta> {
    return arr.reduce(function(acc, v) {
      return acc.concat(v);
    }, [] as Array<AvailableTeamMeta>);
  }

  static getTeams(
    team: string | null, year: string | null, gender: string | null
  ): Array<AvailableTeamMeta> {
    const list = team ?
      (AvailableTeams.byName[team] || ([] as Array<AvailableTeamMeta>)) :
      AvailableTeams.flatten(Object.values(AvailableTeams.byName));

    return list.filter(function(record) {
      return (!team || (team == record.team)) &&
        (!year || (year == record.year)) && (!gender || (gender == record.gender));
    });
  }
  static getTeam(
    team: string, year: string, gender: string
  ): AvailableTeamMeta | null {
      const retVal = AvailableTeams.getTeams(team, year, gender);
      return (retVal.length > 0) ? retVal[0] : null;
  }
  static readonly byName: Record<string, Array<AvailableTeamMeta>> = {
    // Maryland!
    "Maryland": [
      { team: "Maryland", year: "2016/7", gender: "Men", index_template: "maryland" },
      { team: "Maryland", year: "2018/9", gender: "Men", index_template: "misc_conf" },
      { team: "Maryland", year: "2018/9", gender: "Women", index_template: "women_bigten" },
    ],
    // Other BIG
    "Indiana": [
      { team: "Indiana", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Nebraska": [
      { team: "Nebraska", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Wisconsin": [
      { team: "Wisconsin", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Ohio St.": [
      { team: "Ohio St.", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Michigan St.": [
      { team: "Michigan St.", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Iowa": [
      { team: "Iowa", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Michigan": [
      { team: "Michigan", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Purdue": [
      { team: "Purdue", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Illinois": [
      { team: "Illinois", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Penn St.": [
      { team: "Penn St.", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Northwestern": [
      { team: "Northwestern", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Rutgers": [
      { team: "Rutgers", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    "Minnesota": [
      { team: "Minnesota", year: "2018/9", gender: "Men", index_template: "bigten" },
    ],
    // Misc
    "Cincinnati": [
      { team: "Cincinnati", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
    "Kentucky": [
      { team: "Kentucky", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
    "Louisville": [
      { team: "Louisville", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
    "Saint Mary's (CA)": [
      { team: "Saint Mary's (CA)", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
    "Virginia": [
      { team: "Virginia", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
    "UConn": [
      { team: "UConn", year: "2018/9", gender: "Men", index_template: "misc_conf" },
    ],
  }
}
