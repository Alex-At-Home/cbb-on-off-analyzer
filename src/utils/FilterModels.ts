
export class ParamPrefixes {
  static readonly game = "game-";
  static readonly lineup = "lineup-";
}

/** Params for game filtering */
export type GameFilterParams = {
  year?: string,
  team?: string,
  gender?: string,
  autoOffQuery?: string;
  onQuery?: string,
  offQuery?: string,
  baseQuery?: string,
  minRank?: string,
  maxRank?: string
}

/** Params for lineup filtering */
export type LineupFilterParams = {
  year?: string,
  team?: string,
  gender?: string,
  lineupQuery?: string,
  minRank?: string,
  maxRank?: string,
  // For sorting in the generated table:
  minPoss?: string,
  maxTableSize?: string,
  sortBy?: string
}

export class ParamDefaults {
  // Lineup
  static readonly defaultLineupMinPos = "5";
  static readonly defaultLineupMaxTableSize = "50";
  static readonly defaultLineupSortBy = "desc:off_poss";
  // Common
  static readonly defaultTeam = "";
  static readonly defaultYear = "2019/20";
  static readonly defaultGender = "Men";
  static readonly defaultMinRank = "0";
  static readonly defaultMaxRank = "400";
}
