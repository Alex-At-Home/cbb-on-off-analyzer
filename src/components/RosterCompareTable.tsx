// React imports:
import React, { useState } from 'react';

// Next imports:
import { NextPage } from 'next';

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Card from 'react-bootstrap/Card';

// Additional components:
// @ts-ignore
import LoadingOverlay from 'react-loading-overlay';

// Component imports
import GenericTable, { GenericTableOps, GenericTableColProps } from "./GenericTable"

// Util imports
import { CbbColors } from "../utils/CbbColors"

export type RosterCompareModel = {
  on: any,
  off: any,
  baseline: any,
  error_code?: string
}
type Props = {
  rosterCompareStats: RosterCompareModel
}

const RosterCompareTable: React.FunctionComponent<Props> = ({rosterCompareStats}) => {

  const tableFields = { //accessors vs column metadata
    "title": GenericTableOps.addTitle("", ""),
    "sep0": GenericTableOps.addColSeparator(),
    "pct": GenericTableOps.addPctCol("%", "Percentage of possessions played", GenericTableOps.defaultColorPicker),
    "poss": GenericTableOps.addIntCol("Poss", "Number of possessions played", GenericTableOps.defaultColorPicker),
  };
  function tableData(key: string) {
    const obj = (rosterCompareStats as any)[key];
    const total = obj?.global_poss_count?.value || 1;
    return (obj?.player?.buckets || []).map(function(obj: any) {
      return GenericTableOps.buildDataRow({
        title: obj.key,
        pct: 1.0*obj.poss_count.value/total,
        poss: obj.poss_count
      }, GenericTableOps.defaultFormatter, GenericTableOps.defaultCellMeta);
    });
  }
  /** Sticks an overlay on top of the table if no query has ever been loaded */
  function needToLoadQuery() {
    return (Object.keys(rosterCompareStats.on).length == 0) &&
      (Object.keys(rosterCompareStats.off).length == 0) &&
      (Object.keys(rosterCompareStats.baseline).length == 0);
  }

  return <LoadingOverlay
    active={needToLoadQuery()}
    text={rosterCompareStats.error_code ?
      `Query Error: ${rosterCompareStats.error_code}` :
      "Press 'Submit' to view results"
    }
  >
    <Row>
        <Col>
          <Card className="w-100">
          <Card.Body>
          <Card.Title>'On' Roster</Card.Title>
          <GenericTable tableCopyId="rosterOnTable" tableFields={tableFields} tableData={tableData("on")}/>
          </Card.Body>
          </Card>
        </Col>
        <Col>
          <Card className="w-100">
          <Card.Body>
          <Card.Title>'Off' Roster</Card.Title>
          <GenericTable tableCopyId="rosterOffTable" tableFields={tableFields} tableData={tableData("off")}/>
          </Card.Body>
          </Card>
        </Col>
        <Col>
          <Card className="w-100">
          <Card.Body>
          <Card.Title>'Baseline' Roster</Card.Title>
          <GenericTable tableCopyId="rosterBaseTable" tableFields={tableFields} tableData={tableData("baseline")}/>
          </Card.Body>
          </Card>
        </Col>
      </Row>
    </LoadingOverlay>;
}

export default RosterCompareTable;
