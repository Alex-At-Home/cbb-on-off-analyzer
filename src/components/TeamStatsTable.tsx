// React imports:
import React from 'react';

// Next imports:
import { NextPage } from 'next';

// Lodash:
import _ from "lodash";

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

// Additional components:
// @ts-ignore
import LoadingOverlay from 'react-loading-overlay';

// Component imports
import GenericTable, { GenericTableOps, GenericTableColProps } from "./GenericTable"

// Util imports
import { CbbColors } from "../utils/CbbColors"
import { CommonTableDefs } from "../utils/CommonTableDefs"

export type TeamStatsModel = {
  on: any,
  off: any,
  baseline: any,
  error_code?: string
}
type Props = {
  teamStats: TeamStatsModel
}

const TeamStatsTable: React.FunctionComponent<Props> = ({teamStats}) => {

  const offPrefixFn = (key: string) => "off_" + key;
  const offCellMetaFn = (key: string, val: any) => "off";
  const defPrefixFn = (key: string) => "def_" + key;
  const defCellMetaFn = (key: string, val: any) => "def";

  const teamStatsOn = { off_title:  "'On' Offense", def_title: "'On' Defense", ...teamStats.on };
  const teamStatsOff = { off_title:"'Off' Offense", def_title: "'Off' Defense", ...teamStats.off };
  const teamStatsBaseline = { off_title: "'Baseline' Offense", def_title: "'Baseline' Defense", ...teamStats.baseline };

  const tableData = _.flatMap([
    (teamStats.on?.doc_count) ? [
      GenericTableOps.buildDataRow(teamStatsOn, offPrefixFn, offCellMetaFn),
      GenericTableOps.buildDataRow(teamStatsOn, defPrefixFn, defCellMetaFn),
      GenericTableOps.buildRowSeparator()
    ] : [],
    (teamStats.off?.doc_count) ? [
      GenericTableOps.buildDataRow(teamStatsOff, offPrefixFn, offCellMetaFn),
      GenericTableOps.buildDataRow(teamStatsOff, defPrefixFn, defCellMetaFn),
      GenericTableOps.buildRowSeparator()
    ] : [],
    [ GenericTableOps.buildDataRow(teamStatsBaseline, offPrefixFn, offCellMetaFn),
    GenericTableOps.buildDataRow(teamStatsBaseline, defPrefixFn, defCellMetaFn),
    GenericTableOps.buildRowSeparator() ],
  ]);

  // 2] Data Model

  // 3] Utils
  /** Sticks an overlay on top of the table if no query has ever been loaded */
  function needToLoadQuery() {
    return (Object.keys(teamStats.on).length == 0) &&
      (Object.keys(teamStats.off).length == 0) &&
      (Object.keys(teamStats.baseline).length == 0);
  }

  // 4] View
  return <Container>
    <LoadingOverlay
      active={needToLoadQuery()}
      text={teamStats.error_code ?
        `Query Error: ${teamStats.error_code}` :
        "Press 'Submit' to view results"
      }
    >
      <Row>
        <Col>
          <GenericTable tableCopyId="teamStatsTable" tableFields={CommonTableDefs.onOffTable} tableData={tableData}/>
        </Col>
      </Row>
    </LoadingOverlay>
  </Container>;
};

export default TeamStatsTable;
