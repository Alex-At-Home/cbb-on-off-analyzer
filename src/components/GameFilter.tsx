// React imports:
import React, { useState, useEffect } from 'react';

// Next imports:
import { NextPage } from 'next';

// Lodash:
import _ from "lodash";

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

// Component imports:
import { TeamStatsModel } from '../components/TeamStatsTable';
import { RosterCompareModel } from '../components/RosterCompareTable';
import { RosterStatsModel } from '../components/RosterStatsTable';
import { LineupStatsModel } from '../components/LineupStatsTable';
import CommonFilter, { GlobalKeypressManager } from '../components/CommonFilter';
import { ParamPrefixes, FilterParamsType, CommonFilterParams, GameFilterParams, FilterRequestInfo, ParamPrefixesType, ParamDefaults } from "../utils/FilterModels";
import AutoSuggestText from './shared/AutoSuggestText';

// Utils
import { StatModels, OnOffBaselineEnum, OnOffBaselineGlobalEnum, PlayerCode, PlayerId, Statistic, IndivStatSet, TeamStatSet, LineupStatSet } from "../utils/StatModels";
import { QueryUtils } from '../utils/QueryUtils';

type Props = {
  onStats: (teamStats: TeamStatsModel, rosterCompareStats: RosterCompareModel, rosterStats: RosterStatsModel, lineupStats: LineupStatsModel[]) => void;
  startingState: GameFilterParams;
  onChangeState: (newParams: GameFilterParams) => void;
  forceReload1Up: number;
}

const GameFilter: React.FunctionComponent<Props> = ({onStats, startingState, onChangeState, forceReload1Up}) => {

  // Data model

  const {
    // Team stats
    teamDiffs: startTeamDiffs,
    showTeamPlayTypes: startShowTeamPlayTypes,
    showRoster: startShowRoster,
    //(common visualization fields across all tables)
    //(manual overrides)
    manual: startManual,
    showPlayerManual: startShowPlayerManual,
    showOnBallConfig: startShowOnBallConfig,
    //(luck)
    luck: startLuck,
    //(these fields are for the team view)
    onOffLuck: startOnOffLuck,
    showOnOffLuckDiags: startShowOnOffLuckDiags,
    showPlayerOnOffLuckDiags: startShowPlayerOnOffLuckDiags,
    calcRapm: startCalcRapm,
    //(these fields are for the individual view)
    filter: startFilter, sortBy: startSortBy,
    showBase: startShowBase, showExpanded: startShowExpanded,
    showDiag: startShowDiag, possAsPct: startPossAsPct,
    showPosDiag: startShowPosDiag,
    showPlayerPlayTypes: startShowPlayerPlayTypes,
    //these fields affect the query
    autoOffQuery: startAutoOffQuery,
    onQuery: startOnQuery, offQuery: startOffQuery,
    ...startingCommonFilterParams
  } = startingState;

  /** The state managed by the CommonFilter element */
  const [ commonParams, setCommonParams ] = useState(startingCommonFilterParams as CommonFilterParams);

  /** Ugly pattern that is part of support for force reloading */
  const [ internalForceReload1Up, setInternalForceReload1Up ] = useState(forceReload1Up);

  useEffect(() => { // Whenever forceReload1Up is incremented, reset common params:
    if (forceReload1Up != internalForceReload1Up) {
      setCommonParams(startingCommonFilterParams as CommonFilterParams);
      setInternalForceReload1Up(forceReload1Up);
      // Actually have to reset these two vs just their underlying value
      // (could build that intermediate pair,. but we'll stick with this limitation for now)
      setOnQuery(startOnQuery || "");
      setOffQuery(startOffQuery || "");
      //(leave toggleAutoOffQuery since it seems harmless, and weird stuff happened when I tried to set it
      // which I don't have time to investigate):
      //toggleAutoOffQuery(startAutoOffQuery);
    }
  }, [ forceReload1Up ]);

  // Game Filter - custom queries and filters:

  const [ autoOffQuery, toggleAutoOffQuery ] = useState(
    _.isNil(startAutoOffQuery) ? ParamDefaults.defaultAutoOffQuery : startAutoOffQuery
  );
  const [ onQuery, setOnQuery ] = useState(startOnQuery || "");
  const [ offQuery, setOffQuery ] = useState(startOffQuery || "");

  /** Used to differentiate between the different implementations of the CommonFilter */
  const cacheKeyPrefix = ParamPrefixes.game;

  // Utils

  /** Bridge between the callback in CommonFilter and state management */
  function updateCommonParams(params: CommonFilterParams) {
    setCommonParams(params);
  }

  /** Builds a game filter from the various state elements, and also any secondary filters
   * NOTE: ugly hack I need to fix, needs to sync with CommonFilter.onSeeExample
  */
  function buildParamsFromState(includeFilterParams: Boolean): [ GameFilterParams, FilterRequestInfo[] ]  {
    const primaryRequest: GameFilterParams = includeFilterParams ?
      _.assign(
        buildParamsFromState(false)[0], {
          // Team stats
          autoOffQuery: autoOffQuery,
          teamDiffs: startTeamDiffs,
          showTeamPlayTypes: startShowTeamPlayTypes,
          showRoster: startShowRoster,
          // Common luck stats across all tables:
          //(manual overrides)
          manual: startManual,
          showPlayerManual: startShowPlayerManual,
          showOnBallConfig: startShowOnBallConfig,
          //(luck)
          luck: startLuck,
          onOffLuck: startOnOffLuck,
          showOnOffLuckDiags: startShowOnOffLuckDiags,
          showPlayerOnOffLuckDiags: startShowPlayerOnOffLuckDiags,
          // Individual stats:
          calcRapm: startCalcRapm,
          filter: startFilter, sortBy: startSortBy,
          showBase: startShowBase, showExpanded: startShowExpanded,
          showDiag: startShowDiag, possAsPct: startPossAsPct,
          showPosDiag: startShowPosDiag,
          showPlayerPlayTypes: startShowPlayerPlayTypes
      }) : {
        ...commonParams,
        onQuery: onQuery,
        offQuery: offQuery
      };

    //(another ugly hack to be fixed - remove default optional fields)
    QueryUtils.cleanseQuery(primaryRequest);

    const entireSeasonRequest = { // Get the entire season of players for things like luck adjustments
      team: primaryRequest.team, year: primaryRequest.year, gender: primaryRequest.gender,
      minRank: ParamDefaults.defaultMinRank, maxRank: ParamDefaults.defaultMaxRank,
      baseQuery: "", onQuery: "", offQuery: ""
    };
    //TODO: also if the main query minus/on-off matches can't we just re-use that?!
    // (ie and just ignore the on-off portion)

    const alsoPullLineups = (startCalcRapm || startShowRoster);

    const [ baseQuery, maybeAdvBaseQuery ] = alsoPullLineups ?
      QueryUtils.extractAdvancedQuery(commonParams.baseQuery || "") : [ "", undefined ];

    // RAPM calculations:
    const getLineupQuery = (onOrOffQuery: string) => {
      const [ onOrOff, maybeAdvOnOrOff ] = QueryUtils.extractAdvancedQuery(onOrOffQuery);
      const baseToUse = maybeAdvBaseQuery || baseQuery || "";
      const onOffToUse = maybeAdvOnOrOff || onOrOff || "";
      return (baseToUse != "") ? `(${onOffToUse}) AND (${baseToUse})` : onOffToUse;
    };
    const lineupRequests = alsoPullLineups ? [ QueryUtils.cleanseQuery({
      ...commonParams
    }) ].concat((onQuery != "") ? [ QueryUtils.cleanseQuery({
        ...commonParams,
        baseQuery: getLineupQuery(onQuery)
      }) ] : []
    ).concat((offQuery != "") ? [ QueryUtils.cleanseQuery({
        ...commonParams,
        baseQuery: getLineupQuery(offQuery)
      }) ] : []
    ) : [];

    const makeGlobalRequest = !_.isEqual(entireSeasonRequest, primaryRequest);

    return [ primaryRequest, [{
        context: ParamPrefixes.roster as ParamPrefixesType, paramsObj: primaryRequest
      }, {
        context: ParamPrefixes.player as ParamPrefixesType, paramsObj: primaryRequest, includeRoster: !makeGlobalRequest
      }].concat(makeGlobalRequest ? [{ //(don't make a spurious call)
        context: ParamPrefixes.player as ParamPrefixesType, paramsObj: entireSeasonRequest, includeRoster: true
      }] : []).concat(lineupRequests.map(req => {
        return { context: ParamPrefixes.lineup as ParamPrefixesType, paramsObj: req };
      }))
    ];
  }

  /** Handles the response from ES to a stats calc request */
  function handleResponse(jsonResps: any[], wasError: Boolean) {
    const jsonStatuses = jsonResps.map(j => j.status);
    const teamJson = jsonResps?.[0]?.responses?.[0] || {}; //(from primary request)
    const rosterCompareJson = jsonResps?.[1]?.responses?.[0] || {}; //(from roster request)
    const rosterStatsJson = jsonResps?.[2]?.responses?.[0] || {}; //(from player request #1)

    // 3, [4, 5] can be lineups ... or they might be 4, [5, 6]
    // depends on whether jsonResps?.[3]?.responses?.[0] has "aggregations.tri_filter"

    //(optionally, from player request #2)
    const hasGlobalRosterStats = jsonResps?.[3]?.responses?.[0]?.aggregations?.tri_filter;
    const globalRosterStatsJson =
      (hasGlobalRosterStats ? jsonResps?.[3]?.responses?.[0] : undefined) || _.cloneDeep(rosterStatsJson);
      //(need to clone it so that changes to baseline don't overwrite global)

    const globalTeam = teamJson?.aggregations?.global?.only?.buckets?.team || StatModels.emptyTeam();
    const rosterInfo = jsonResps?.[hasGlobalRosterStats ? 3 : 2]?.roster;
    if (rosterInfo) {
      globalTeam.roster = rosterInfo;
    }

    /** For RAPM, from lineup requests */
    const lineupResponses = _.drop(jsonResps, hasGlobalRosterStats ? 4 : 3).map(lineupJson => {
      return {
        lineups: lineupJson?.responses?.[0]?.aggregations?.lineups?.buckets,
        error_code: wasError ? (lineupJson?.status || jsonStatuses?.[0] || "Unknown") : undefined
      };
    });

    onStats({
      on: teamJson?.aggregations?.tri_filter?.buckets?.on || StatModels.emptyTeam(),
      off: teamJson?.aggregations?.tri_filter?.buckets?.off || StatModels.emptyTeam(),
      onOffMode: autoOffQuery,
      baseline: teamJson?.aggregations?.tri_filter?.buckets?.baseline || StatModels.emptyTeam(),
      global: globalTeam,
      error_code: wasError ? (teamJson?.status || jsonStatuses?.[0] || "Unknown") : undefined
    }, {
      on: rosterCompareJson?.aggregations?.tri_filter?.buckets?.on || {},
      off: rosterCompareJson?.aggregations?.tri_filter?.buckets?.off || {},
      onOffMode: autoOffQuery,
      baseline: rosterCompareJson?.aggregations?.tri_filter?.buckets?.baseline || {},
      error_code: wasError ? (rosterCompareJson?.status || jsonStatuses?.[1] || "Unknown") : undefined
    }, {
      on: rosterStatsJson?.aggregations?.tri_filter?.buckets?.on?.player?.buckets || [],
      off: rosterStatsJson?.aggregations?.tri_filter?.buckets?.off?.player?.buckets || [],
      onOffMode: autoOffQuery,
      baseline: rosterStatsJson?.aggregations?.tri_filter?.buckets?.baseline?.player?.buckets || [],
      global: globalRosterStatsJson?.aggregations?.tri_filter?.buckets?.baseline?.player?.buckets || [],
      error_code: wasError ?
        (rosterStatsJson?.status || jsonStatuses?.[2] ||
          globalRosterStatsJson?.status || jsonStatuses?.[3] || "Unknown") : undefined
    }, lineupResponses);
  }

  /** Sets the automatically generated off query, if that option is selected */
  const setAutoOffQuery = (onQuery: string) => {
    setOffQuery((onQuery == "") || (onQuery == " ") ? "" : `NOT (${onQuery})`);
  }

  /** Ran into issues with SSR and 'readOnly' property, so have to fix like this */
  function renderOffQueryFormField() {
    if (typeof window !== `undefined`) {
      return <Form.Control
        placeholder="eg 'NOT (Player1 AND (Player2 OR Player3))'"
        onKeyUp={(ev: any) => setOffQuery(ev.target.value)}
        onChange={(ev: any) => setOffQuery(ev.target.value)}
        value={offQuery}
        readOnly={autoOffQuery}
      />
    }
  }

  /** Works around a bug in the input where it was ignoring the first select/delete of a page load */
  const handleOnQueryChange = (ev: any) => {
    setOnQuery(ev.target.value);
    if (autoOffQuery) {
      setAutoOffQuery(ev.target.value);
    }
  };

  const maybeOn = autoOffQuery ? "On ('A')" : "'A'";
  const maybeOff = autoOffQuery ? "Off ('B')" : "'B'";

  // Visual components:

  return <CommonFilter //(generic type inferred)
      startingState={startingState}
      onChangeState={onChangeState}
      onChangeCommonState={updateCommonParams}
      tablePrefix = {cacheKeyPrefix}
      buildParamsFromState={buildParamsFromState}
      childHandleResponse={handleResponse}
      forceReload1Up={internalForceReload1Up}
    ><GlobalKeypressManager.Consumer>{ globalKeypressHandler => <div>
      <Form.Group as={Row}>
        <Form.Label column sm="2">{maybeOn} Query</Form.Label>
        <Col sm="8">
          <AutoSuggestText
            readOnly={false}
            placeholder="eg 'Player1 AND (Player2 OR Player3)'"
            initValue={onQuery}
            year={commonParams.year}
            gender={commonParams.gender}
            team={commonParams.team}
            onKeyUp={handleOnQueryChange}
            onChange={handleOnQueryChange}
            onKeyDown={globalKeypressHandler}
          />
        </Col>
      </Form.Group>
      <Form.Group as={Row}>
        <Form.Label column sm="2">{maybeOff} Query</Form.Label>
        <Col sm="8">
          { (typeof window !== `undefined`) ?
            <AutoSuggestText
              readOnly={autoOffQuery}
              placeholder="eg 'NOT (Player1 AND (Player2 OR Player3))'"
              initValue={offQuery}
              year={commonParams.year}
              gender={commonParams.gender}
              team={commonParams.team}
              onKeyUp={(ev: any) => setOffQuery(ev.target.value)}
              onChange={(ev: any) => setOffQuery(ev.target.value)}
              onKeyDown={globalKeypressHandler}
            /> : <div/> //(this construct needed to address SSR/readonly issue)
          }
        </Col>
        <Col sm="2">
          <Form.Check type="switch"
            id="autoOffQuery"
            checked={autoOffQuery}
            onChange={() => {
              if (!autoOffQuery) {
                setAutoOffQuery(onQuery);
              }
              toggleAutoOffQuery(!autoOffQuery);
            }}
            label="Auto"
          />
        </Col>
      </Form.Group>
    </div>}</GlobalKeypressManager.Consumer></CommonFilter>
    ;
}

export default GameFilter;
