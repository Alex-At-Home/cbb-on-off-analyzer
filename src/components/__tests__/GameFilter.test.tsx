import React from "react";
import GameFilter from "../GameFilter";
import { GameFilterParams } from "../../utils/FilterModels";
import { TeamStatsModel } from "../TeamStatsTable";
import { RosterCompareModel } from "../RosterCompareTable";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

describe("GameFilter", () => {
  test("GameFilter - should create snapshot", () => {
    const dummySubmitCallback = (
      teamStats: TeamStatsModel,
      rosterCompareStats: RosterCompareModel
    ) => {};
    const dummyChangeStateCallback = (stats: GameFilterParams) => {};
    const wrapper = shallow(
      <GameFilter
        onStats={dummySubmitCallback}
        startingState={{}}
        onChangeState={dummyChangeStateCallback}
        forceReload1Up={0}
        testMode={true}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
//TODO: add mock for fetch and test submit etc
//TODO: check text
