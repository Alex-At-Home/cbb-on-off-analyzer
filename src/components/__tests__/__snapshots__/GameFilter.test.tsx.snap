// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameFilter GameFilter - should create snapshot 1`] = `
<CommonFilter
  buildParamsFromState={[Function]}
  childHandleResponse={[Function]}
  childSubmitRequest={[Function]}
  onChangeCommonState={[Function]}
  onChangeState={[Function]}
  startingState={Object {}}
  tablePrefix="game-"
>
  <FormGroup
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "noGutters": false,
        },
        "render": [Function],
      }
    }
  >
    <FormLabel
      column={true}
      sm="2"
      srOnly={false}
    >
      On Query
    </FormLabel>
    <Col
      sm="8"
    >
      <FormControl
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="eg 'Player1 AND (Player2 OR Player3)'"
        value=""
      />
    </Col>
  </FormGroup>
  <FormGroup
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "noGutters": false,
        },
        "render": [Function],
      }
    }
  >
    <FormLabel
      column={true}
      sm="2"
      srOnly={false}
    >
      Off Query
    </FormLabel>
    <Col
      sm="8"
    />
    <Col
      sm="2"
    >
      <FormCheck
        checked={true}
        disabled={false}
        id="autoOffQuery"
        inline={false}
        isInvalid={false}
        isValid={false}
        label="Auto"
        onChange={[Function]}
        title=""
        type="switch"
      />
    </Col>
  </FormGroup>
  <FormGroup
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "noGutters": false,
        },
        "render": [Function],
      }
    }
  >
    <FormLabel
      column={true}
      sm="2"
      srOnly={false}
    >
      Baseline Query
    </FormLabel>
    <Col
      sm="8"
    >
      <FormControl
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="eg 'NOT (WalkOn1 OR WalkOn2)' - applied to both 'On' and 'Off' queries"
        value=""
      />
    </Col>
    <Col
      sm="2"
    >
      <FormCheck
        checked={false}
        disabled={true}
        id="excludeWalkons"
        inline={false}
        isInvalid={false}
        isValid={false}
        label="Auto Walk-ons"
        title=""
        type="switch"
      />
    </Col>
  </FormGroup>
</CommonFilter>
`;
